nixos/modules/config/shells-environment.nix

```diff
@@ -210,14 +210,10 @@ in
         ''}
       '';
 
-    system.activationScripts.binsh = stringAfter [ "stdio" ]
-      ''
-        # Create the required /bin/sh symlink; otherwise lots of things
-        # (notably the system() function) won't work.
-        mkdir -m 0755 -p /bin
-        ln -sfn "${cfg.binsh}" /bin/.sh.tmp
-        mv /bin/.sh.tmp /bin/sh # atomically replace /bin/sh
-      '';
+    systemd.tmpfiles.rules = [
+      "d /bin 0755 -"
+      "L+ /bin/sh - - - - ${cfg.binsh}"
```

Comment:
isn't this too late? `/bin/sh` could be used well before `systemd-tmpfiles-setup.service` runs. Same goes for `/usr/bin/env`, though to a lesser degree since `/bin/sh` is used by `system(3)`.

Reply:
I hostely don't know when this is needed. I ran a bunch of tests and they all seemed to be fine. tmpfiles runs before sysinit.target and this is really the earliest we can order anything other than already running it in the initrd.

Reply:
It's needed by anything that uses `system(3)`, at least. That may not include anything in `systemd`, but who knows what else is out there. I think assuming `/bin/sh` and `/usr/bin/env` is a *bare minimum* that application developers expect of the host system, and those applications may as well be run during early boot.

Reply:
We actually cannot move this into the initrd because this would break initrd-less systems (e.g. containers).  

Normal services are started after basic.target, by which the links already have been established. You would only need /bin/sh earlier if you explicitly order your service earlier. Overall I think its fair to say that the system is not up before basic.target and you shouldn't expect the existence of much. For example you might not even have /tmp or /var/tmp before that, so why expect /bin/sh?

Reply:
But units ordered before `basic.target` are entirely reasonable, and very well could require `/bin/sh`. The difference between `/bin/sh` and `/tmp` or `/var/tmp` is that `/bin/sh` *has always existed at all times in the system forever*, ***and*** a fundamental system call `system(3)` relies on it.

Reply:
Of course NixOS has never respected FHS, but there's a reason that an exception was made for `/bin/sh` and `/usr/bin/env`, and I think breaking that would be a bridge too far

Reply:
> systemd supports hermetic /usr

But you're failing that as well. `/usr/bin/env` in this PR is also ordered far too late.

It's a good point about `/bin/sh` though.